Paper: Python Array API Standard: Toward Array Interoperability in the Scientific Python Ecosystem (#822)
scipy-conference/scipy_proceedings: papers/aaron_meurer/paper.rst

:email: kgryte@gmail.com
:equal-contributor:
:institution: Quansight

:author: Ralf Gommers
:email: ralf.gommers@gmail.com
:equal-contributor:
:institution: Quansight

:author: Yao-Lung L. Fang
:email: leof@nvidia.com
:equal-contributor:
:institution: NVIDIA Corporation

:author: Matthew Barber
:email: quitesimplymatt@gmail.com
:equal-contributor:
:institution: Quansight

:author: Stephan Hoyer
:email: shoyer@google.com
:institution: Google

:author: Andreas Müller
:email: amueller@microsoft.com
:institution: Microsoft

:author: Sheng Zha
:email: zhasheng@apache.org
:institution: Amazon

:author: Saul Shanabrook
:email: s.shanabrook@gmail.com

:author: Stephannie Jiménez Gacha
:email: sgacha@quansight.com
:institution: Quansight

:author: Mario Lezcano-Casado
:email: mlezcano@quansight.com
:institution: Quansight

:author: Thomas J. Fan
:email: thomasjpfan@gmail.com
:institution: Quansight

:author: Tyler Reddy
:email: treddy@lanl.gov
:institution: LANL

:author: Alexandre Passos
:email: alexandre.tp@gmail.com

:author: Hyukjin Kwon
:email: gurwls223@apache.org
:institution: Databricks

:author: Travis E. Oliphant
:email: travis@quansight.com
:institution: Quansight

> 

Reviewer: ```suggestion

:author: John Kirkham
:email: jkirkham@nvidia.com
:equal-contributor:
:institution: NVIDIA Corporation

```

xref: https://github.com/data-apis/scipy-2023-presentation/pull/25

:author: Consortium for Python Data API Standards
:email:

:bibliography: bibliography

.. Note: treat the Consortium as being equivalent to a PI (i.e., list it last without explicit equal contribution)

===========================================================================================
Python Array API Standard: Toward Array Interoperability in the Scientific Python Ecosystem
===========================================================================================

.. class:: abstract

   The Python array API standard specifies standardized application programming
   interfaces (APIs) and behaviors for array and tensor objects and operations
   as commonly found in libraries such as NumPy :cite:`Harris2020a`, PyTorch
   :cite:`Paszke2019a`, TensorFlow :cite:`Abadi2016a`, Dask :cite:`Rocklin2015a`,

>    and CuPy :cite:`Okuta2017a`. The establishment and subsequent adoption of the

Reviewer 2: Here's a copy-edit thing: there are several lists of array library names in the paper. They are necessarily different lists, but they could at least be put in a common order to make it easier to compare them.

The section highlighted above, in the abstract, are just examples of libraries. There probably wasn't any criterion for what is included or excluded other than likelihood that a reader of the paper would be familiar with them.

The line

https://github.com/scipy-conference/scipy_proceedings/blob/98ebaee556e0c3c4c78d580e072047db0a3eee77/papers/aaron_meurer/paper.rst?plain=1#L323-L324

has libraries that were consulted as prior art in the design phase.

The line

https://github.com/scipy-conference/scipy_proceedings/blob/98ebaee556e0c3c4c78d580e072047db0a3eee77/papers/aaron_meurer/paper.rst?plain=1#L744-L745

has libraries in the compatibility layer.

The line

https://github.com/scipy-conference/scipy_proceedings/blob/98ebaee556e0c3c4c78d580e072047db0a3eee77/papers/aaron_meurer/paper.rst?plain=1#L679-L681

has libraries that are in the process of adopting the standard.

Fig. 2 and discussion has libraries used in the scikit-learn and SciPy tests. A plot is harder to change than a list order in text; the plot's order is PyTorch followed by CuPy.

Reviewer 2: Maybe also the Index Terms (mostly a list of library names).

Reviewer 3: This has now been addressed. We've updated the content to use a consistent ordering for referenced libraries.

package by porting Numarray's features to Numeric :cite:`Harris2020a`. This
effort was largely successful and resolved the fragmentation at the time. For
roughly a decade, NumPy was the only widely used array library. Building on
NumPy, pandas was subsequently introduced in 2008 in order to address the need
for a high performance, flexible tool for performing quantitative analysis on
labeled tabular data :cite:`McKinney2011a`.

Over the past 10 years, the rise of deep learning and the emergence of new
hardware has led to a proliferation of new libraries and a corresponding
fragmentation within the PyData array and dataframe ecosystem. These libraries
often borrowed concepts from, or entirely copied, the APIs of older libraries,
such as NumPy, and then modified and evolved those APIs to address new needs
and use cases. Although the communities of each library individually discussed
interchange and interoperability, no general coordination arose among libraries
to avoid further fragmentation and to arrive at a common set of API standards
until the founding of the Consortium.

The genesis for the Consortium grew out of many conversations among maintainers

> during 2019-2020. During those conversations, it quickly became clear that any

Reviewer 2: ```suggestion
during 2019‒2020. During those conversations, it quickly became clear that any
```

(n-dash)

Reviewer 3: This has been addressed. Thanks for catching!

Objectives
----------

Standardization efforts must maintain a delicate balance between codifying what
already exists and maintaining relevance with respect to future innovation. The
latter aspect is particularly fraught, as relevance requires anticipating
future needs, technological advances, and emerging use cases. Accordingly, if a
standard is to remain relevant, the standardization process must be
conservative in its scope, thorough in its consideration of current and prior
art, and have clearly defined objectives against which continued success is
measured.

To this end, we established four objectives for the array API standard. 1)
Allow array-consuming libraries to accept and operate on arrays
from multiple different array libraries. 2) Establish a common set of
standardized APIs and behaviors, enabling more sharing and code reuse. 3) For

> new array libraries, offer a concrete API that can be adopted as is. 4)

Reviewer 2: ```suggestion
new array libraries, offer a concrete API that can be adopted as-is. 4)
```

Reviewer 3: This has been addressed. Thanks for catching!

consumers are likely to need to modify code in order to ensure optimal
performance and behavior. 3) Support mixing multiple array libraries in a
single function call. Mixing array libraries requires defining hierarchies and
specifying rules for device synchronization and data localization. Such rules
are likely to be specific to individual use cases.

Design Principles
-----------------

In order to define the contours of the standardization process, we established
the following design principles:

**Pure functions.** The standardized API should consist primarily of
functions. Functional API design is the dominant pattern among array
libraries, both in Python and in other frequently used programming languages
supporting array computation, such as MATLAB :cite:`Moler2020a` and Julia
:cite:`Bezanson2017a`. While method chaining and the fluent interface design
pattern are also relatively common, especially among array libraries supporting
deferred execution and operator fusion, functional APIs are generally preferred.
This mirrors design patterns used in underlying implementations, such as those
written in C/C++ and Fortran, and more closely matches written mathematical

> notation.

Reviewer 2: I found this use of the words "pure functions" and "functional API" to be confusing. As per [Wikipedia](https://en.wikipedia.org/wiki/Functional_programming):

> Functional programming is sometimes treated as synonymous with [purely functional programming](https://en.wikipedia.org/wiki/Purely_functional_programming), a subset of functional programming which treats all functions as [deterministic](https://en.wikipedia.org/wiki/Deterministic_system) mathematical [functions](https://en.wikipedia.org/wiki/Function_(mathematics)), or [pure functions](https://en.wikipedia.org/wiki/Pure_function).

I was expecting these words to mean that the API is designed to be side-effect free, but the text of this paragraph suggests that you mean free-standing functions, as opposed to methods (or binary operators?).

This point would be clearer if you replace the words "pure functions" and "functional API" with something else, maybe "unbound functions, rather than methods".

Reviewer 3: This has now been addressed. We replaced "functional" and "pure functions" with "standalone functions".

**Minimal array object.** The standard should not require that an array object
have any attributes or methods beyond what is necessary for inspection
(e.g., shape, data type, and device) or for supporting operator overloading
(i.e., magic methods). [#]_

.. [#] Notably, array strides should be considered an implementation detail and should not be required as a public Python attribute.

**No dependencies.** The standard and its implementations should not require
any dependencies outside of Python itself.

**Accelerator support.** Standardized APIs and behaviors should be possible to
implement for both central processing units (CPUs) and hardware-accelerated
devices, such as graphics processing units (GPUs), tensor processing units
(TPUs), and field-programmable gate arrays (FPGAs).

**Compiler support.** Standardized APIs and behaviors should be amenable to
just-in-time (JIT) and ahead-of-time (AOT) compilation and graph-based
optimization techniques, such as those used by PyTorch :cite:`Paszke2019a`, JAX
:cite:`Bradbury2018a`, and TensorFlow :cite:`Abadi2016a`. APIs and behaviors
not amenable to compilation, such as APIs returning arrays having

> data-dependent output shapes, should be specified as optional. Moreover,

Reviewer 2: It's not, strictly speaking, compiler support that's at issue here—the issue is that shapes (just `ndim` or the whole `shape` tuple) are considered part of the data type that gets compiled in, right?

What you want to allow for is libraries that use compile-time literal shapes to more effectively vectorize and unroll loops (and maybe more advanced graph optimization techniques than I know about). For example, Numba knows `ndim` but not `shape` at compile-time and therefore is able to return data-dependent output shapes, but JAX knows the `shape` at compile-time and therefore is not able to (but compiles to faster code).

If so, then maybe it would be better to say here that you want to allow compilers to have any amount of array metadata fields at compile-time. Later, you define an array object's metadata fields as "data type," "shape," and "device," and that's the maximum amount of information you want to allow compilers to have.

Reviewer 4: The "compiler support" point is broader than shapes. Other examples include avoiding functions that are polymorphic (e.g. `numpy.unique` has a number of return values that depend on input keywords) or enforcing output dtype stability (no value-based casting allowed).

> Later, you define an array object's metadata fields as "data type," "shape," and "device," and that's the maximum amount of information you want to allow compilers to have.

Yes, that sounds right for arrays that get returned.

To address this comment, how about adding one or both examples above for compiler support explicitly, and then adding explicitly that dtype, shape and device of any returned arrays should be predictable from the input types/dtypes/shapes/devices only?

Reviewer 2: That sounds like a good solution to me. The comment was about my wrong guess about what "compiler support" entails; illustrating with examples is a good way to do it.

Reviewer 3: This has now been addressed. We added the example of "return type polymorphism" and a sentence on return value predictability based on input arguments.

standardization.

**Distributed support.** Standardized APIs and behaviors should be amenable to
implementation in array libraries supporting distributed computing (e.g., Dask :cite:`Rocklin2015a`).

**Consistency.** Except in scenarios involving backward compatibility concerns,
naming conventions and design patterns should be consistent across
standardized APIs.

**Extensibility.** Conforming array libraries may implement functionality
which is not included in the array API standard. As a consequence, array
consumers bear responsibility for ensuring that their API usage is portable
across specification-conforming array libraries.

**Deference.** Where possible, the array API standard should defer to existing,
widely-used standards. For example, the accuracy and precision of numerical
functions should not be specified beyond the guidance included in IEEE 754 :cite:`IEEE754`.

**Universality.** Standardized APIs and behaviors should reflect common usage
among a wide range of existing array libraries. Accordingly, with rare
exception, only APIs and behaviors having existing implementations and broad
support within the ecosystem may be considered candidates for standardization.

> Methods
> =======

Reviewer 2: Perhaps

```suggestion
Methodology
===========
```

because "methods" are part of an API, and that's not what you mean here.

Reviewer 3: This has now been addressed. Thanks for the suggestion!

based on prevalence and ease of implementation. The common API subset included
function names, method names, attribute names, and positional and keyword
arguments. As an example of a derived API, consider the common API for
computing the arithmetic mean:

.. code:: python

   mean(a, axis=None, keepdims=False)

To assist in determining standardization prioritization, we leveraged usage
data (discussed below) to confirm API need and to inform naming conventions,
supported data types, and optional arguments. We have summarized findings and
published tooling :cite:`Consortium2022c` for additional analysis and
exploration, including Jupyter notebooks :cite:`Kluyver2016a`, as public
artifacts available on GitHub.

Usage
-----

To understand usage patterns of array libraries within the ecosystem, we first
identified a representative sample of commonly used Python libraries
("downstream libraries") which consume the aforementioned array libraries. The
sample of downstream libraries included SciPy :cite:`Virtanen2020a`, pandas
:cite:`McKinney2011a`, Matplotlib :cite:`Hunter2007a`, xarray
:cite:`Hoyer2017a`, scikit-learn :cite:`Pedregosa2011a`, statsmodels

> :cite:`Seabold2010a`, and scikit-image :cite:`Vanderwalt2014a`, among others.

Reviewer 2: This is not a major point, but it seems like it would be possible to publish a complete list of the downstream libraries used to establish usage. You're being very formal with the voting systems that treat their usage fairly (Dowdall and Borda, referenced below), maybe there should also be a complete list of the constituency somewhere?

How large is it? A lot more than these 7 or just enough more that you don't want to use up page space by writing them all?

Reviewer 5: [Here](https://github.com/data-apis/python-record-api/tree/master/data/api) is a list of all the libraries used to create usage data I believe, unless some were added/removed outside of the recording repo.

Author: As Saul noted, the full list of libraries is 

```
alphalens
hvplot
modin
pandas
skimage
xarray
dask
koalas
networkx
prophet
scipy
sklearn
geopandas
matplotlib
orange3
pyjanitor
seaborn
statsmodels
```

which is 18 libraries. We primarily only listed the most widely used libraries here for the sake of space. Should we remove "among others" here and replace it with a full list?

Also, hopefully it is clear that even though we used a formal voting system in this data analysis, this was only used to assign priorities for standardization. At the end of the day the final selection of which APIs to include was made by the Consortium.

Reviewer 3: As mentioned in other comments, we explicitly listed the most widely used libraries in the interest of space. We link (via the bibliography) to the Python record API repository where a more complete list can be found.

An array device specifies the location of array memory allocation. A
conforming array object is assigned to a single logical device. To support
array libraries supporting execution on different device types (e.g., CPUs,
GPUs, TPUs, etc.), conforming libraries must provide standardized device APIs
in order to coordinate execution location. The following example demonstrates
how an array-consuming library might use standardized device APIs to ensure
execution occurs on the same device as the input.

.. code:: python

   def some_function(x):
       # Retrieve a standard-compliant namespace
       xp = x.__array_namespace__()

       # Allocate a new array on the same device as x
       y = xp.linspace(0, 2*xp.pi, 100, device=x.device)

       # Perform computation (on device)
       return xp.sin(y) * x

To interact with array objects, one uses "indexing" to access sub-arrays and
individual elements, "operators" to perform logical and arithmetic operations
(e.g., `+`, `-`, `*`, `/`, and `@`), and array-aware functions (e.g., for
linear algebra, statistical reductions, and element-wise computation). Array
indexing semantics extend built-in Python sequence `__getitem__()` indexing

> semantics to support element access across multiple dimensions (`Fig. 1b`_).

Reviewer 2: Are in-place mutation semantics included? That is, `__setitem__`?

At one point, I thought that in-place mutation was going to be an optional extension, since some array libraries (e.g. JAX again) benefit by excluding it.

Maybe you really did mean "immutable data"/"no side effects" in the "pure functions"/"functional API" part of the Design Principles section—in which case, I misunderstood it!

Author: In-place semantics are supported, but they are only supported in situations where mutations do not affect more than one array (i.e., the behavior is undefined in the presence of "views"). See https://data-apis.org/array-api/latest/design_topics/copies_views_and_mutation.html#copyview-mutability. There's actually a lengthy discussion about this happening at https://github.com/data-apis/array-api/issues/609 if you are interested.

Author: In-place semantics are supported, but they are only supported in situations where mutations affect more than one array (i.e., the behavior is undefined in the presence of "views"). See https://data-apis.org/array-api/latest/design_topics/copies_views_and_mutation.html#copyview-mutability. There's actually a lengthy discussion about this happening at https://github.com/data-apis/array-api/issues/609 if you are interested.

Reviewer 4: > they are only supported in situations where mutations affect more than one array

Correction: only when _only one array_ is affected. I.e., at that point there is no difference between in-place and out-of-place versions of the same operation. Hence there is no issue with libraries implemented under the hood as immutable data structures (JAX, TensorFlow).

Reviewer 3: This has now been addressed. We added a footnote concerning in-place operations.

Indexing an array using a boolean array (also known as "masking") is an
optional standardized behavior. The result of a mask operation is
data-dependent and thus difficult to implement in array libraries relying on

> static analysis for graph-based optimization.

Reviewer 2: This is referring to the "compiler support" part of the Design Principles section: the fact that the `shape` array metadata must be known at compile-time.

That makes sense for indexing using a boolean array, but what about indexing using an integer array, which is more powerful and it does not require shape information to be derived from data? That is, in

```python
output_array = base_array[integer_array]
```

the `output_array.shape == integer_array.shape`.

I'm not saying that the standard needs this; I'm saying that the paper has left me wondering about it, since I usually think of slicing by boolean array and slicing by integer array as going hand-in-hand. (They're presented side-by-side in the same [section of the NumPy documentation](https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing).) A short sentence saying that integer array indexing is included or why it's not would be enough.

Author: Integer array indexing wasn't omitted because of compiler support, but rather because of lack of broad library support, and because of caution surrounding implementing the full NumPy indexing behavior, which includes some poor design decisions that we don't want to encode into the spec. It is possible that a future version of the standard will include some subset of integer array indexing. 

There have been discussions about adding equivalent functional APIs, and the 2022.12 version of the spec includes the `take` function, which implements the equivalent of integer array getitem indexing for 1D arrays. The setitem equivalent, `put`, is still being discussed. 

See https://github.com/data-apis/array-api/issues/177 for the full discussions here.

Reviewer 3: This has now been addressed. We added a note concerning integer array indexing and the possibility of its inclusion in a future revision of the array API standard.

line.

.. code:: python

    def z_score(x):
        return (x - xp.mean(x)) / xp.stdev(x)

In addition to vectorized operations, the array API standard includes, but is
not limited to, functions for creating new arrays, with support for explicit
device allocation, reshaping and manipulating existing arrays, performing
statistical reductions across one, multiple, or all array axes (`Fig. 1e`_), and
sorting array elements. Altogether, these APIs provide a robust and portable
foundation for higher-order array operations and general array computation.

Optional Extensions
-------------------

While a set of commonly used array-aware functions is sufficient for many
array computation use cases, additional, more specialized, functionality may be
warranted. For example, while most data visualization libraries are unlikely to
explicitly rely on APIs for computing Fourier transforms, signal analysis
libraries supporting spectral analysis of time series are likely to require
Fourier transform APIs. To accommodate specialized APIs, the Python array API
standard includes standardized optional extensions.

An extension is defined as a coherent set of standardized functionality which

> is commonly implemented across many, but not all, array libraries. Due to

Reviewer 2: The format of an extension is not limited to a set of functions (e.g. the Fourier analysis functions in your motivating example), so it _could_ be something like defining `strides` as an array metadata field and what happens to the strides in indexing and other functions, right?

But then, I suppose that this extension would need to have something to say about all other extensions: how do strides interact with Fourier functions? That would make it a special extension, and if there are too many special extensions, all commenting on each other, that could become infeasible.

Does the definition of an extension exclude that possibility? Like, if an extension can only be an additional suite of functions, then this sentence could clarify that.

Author: Right now an "extension" in the standard is defined specifically as a sub-namespace https://data-apis.org/array-api/latest/extensions/index.html. This is distinct from "optional behavior" such as boolean indexing. Will make this clearer.

Reviewer 3: This has now been addressed. We clarified that an extension is a sub-namespace of related functionality.

(e.g., boolean to integer, integer to floating-point, etc.) are not specified.
To ensure portability, we must explicitly convert a boolean array to an integer
array before calling `xp.sum()`.

To test the performance implications of refactoring scikit-learn's LDA
implementation, we first generated a random two-class classification problem
having 400,000 samples and 300 features. [#]_ We next devised two benchmarks, one
for fitting an LDA model and the second for predicting class labels for each
simulated sample. We then ran the benchmarks and measured execution time for
NumPy, PyTorch, and CuPy backends on Intel i9-9900K and NVIDIA RTX 2080
hardware. For PyTorch, we collected timings for both CPU and GPU execution
models. To ensure timing reproducibility and reduce timing noise, we repeated
each benchmark ten times and computed the average execution time.

.. [#] To ensure that observed performance is not an artifact of the generated dataset, we tested performance across multiple random datasets and did not observe a measurable difference across benchmark runs.

`Fig. 2a`_ and `Fig. 2b`_ display results, showing average execution time

> relative to NumPy. When fitting an LDA model (`Fig. 2a`_), we observe 1.9x
> higher throughput for PyTorch CPU, 7.9x for CuPy, and 45.1x for PyTorch GPU.
> When predicting class labels (`Fig. 2b`_), we observe 2.5x higher throughput
> for PyTorch CPU, 24.6x for CuPy, and 44.9x for PyTorch GPU. In both benchmarks,

Reviewer 2: ```suggestion
relative to NumPy. When fitting an LDA model (`Fig. 2a`_), we observe 1.9×
higher throughput for PyTorch CPU, 7.9× for CuPy, and 45.1× for PyTorch GPU.
When predicting class labels (`Fig. 2b`_), we observe 2.5× higher throughput
for PyTorch CPU, 24.6× for CuPy, and 44.9× for PyTorch GPU. In both benchmarks,
```

Reviewer 3: This has been addressed. Thanks!

signal processing, and image processing, among others. Similar to scikit-learn,
its current implementation relies heavily on NumPy. We thus sought to test
whether SciPy could benefit from adopting the Python array API standard.

Following a similar approach to the scikit-learn benchmarks, we identified
SciPy's signal processing APIs as being amenable to input arrays supporting
alternative execution models and selected an API for estimating the power
spectral density using Welch's method :cite:`Welch1967a` as a representative
test case. We then generated a representative synthetic test signal (a 2 Vrms
sine wave at 1234 Hz, corrupted by 0.001 :math:`\text{V}^2/\text{Hz}` of white
noise sampled at 10 kHz) having 50,000,000 data points. We next devised two
benchmarks, one using library-specific optimizations and a second strictly
using APIs in the array API standard. We ran the benchmarks for the same
backends, on the same hardware, and using the same analysis approach as the
scikit-learn benchmarks discussed above.

`Fig. 2c`_ and `Fig. 2d`_ display results, showing average execution time
relative to NumPy. When using library-specific optimizations (`Fig. 2c`_), we

> observe 1.4x higher throughput for PyTorch CPU, 51.6x for PyTorch GPU, and
> 52.4x for CuPy. When omitting library-specific optimizations (`Fig. 2d`_), we
> observe a 12-25x **decreased** throughput across all non-NumPy backends.

Reviewer 2: ```suggestion
observe 1.4× higher throughput for PyTorch CPU, 51.6× for PyTorch GPU, and
52.4× for CuPy. When omitting library-specific optimizations (`Fig. 2d`_), we
observe a 12‒25× **decreased** throughput across all non-NumPy backends.
```

Reviewer 3: This has now been addressed. Thanks!

we expect to continue working with array and array-consuming library
maintainers to identify and codify APIs and behaviors suitable for
standardization.

**Adoption**: In order to ensure the continued success and relevance of the
Python array API standard, we work closely with maintainers of array and
array-consuming libraries to facilitate adoption by soliciting feedback,
addressing pain points, and resolving any specification ambiguities. In the
immediate future, we plan to release additional tooling for tracking adoption
and measuring specification compliance. For the former, we are collecting
static compliance data and will publish compatibility tables as part of the
array API standard publicly available on-line. For the latter, we are
developing an automated test suite reporting system to gather array API test
suite results from array libraries as part of their continuous integration
pipelines. We expect these tools to be particularly valuable to array-consuming
libraries in order to quickly assess API portability.

**Coordination**: Providing a forum for coordination among array libraries
(and their consumers) was the primary motivating factor behind Consortium

> formation and is the most important byproduct of Consortium efforts. By

Reviewer 2: It seems like this would be a good place to point readers to your communications channels. Is it all on https://github.com/data-apis/array-api, or are there Discourse/Discord sites for people to join?

Reviewer 3: This has now been addressed. We added a note referring readers to the public issue tracker.

(e.g., `+`, `-`, `*`, `/`, and `@`), and array-aware functions (e.g., for
linear algebra, statistical reductions, and element-wise computation). Array
indexing semantics extend built-in Python sequence `__getitem__()` indexing
semantics to support element access across multiple dimensions (`Fig. 1b`_). [#]_
Indexing an array using a boolean array (also known as "masking") is an
optional standardized behavior. [#]_ The result of a mask operation is
data-dependent and thus difficult to implement in array libraries relying on
static analysis for graph-based optimization.

> .. [#] The array API standard includes support for in-place operations via `__setitem__()`; however, behavior is undefined if an in-place operation would affect arrays other than the target array (e.g., in array libraries supporting multiple "views" of the same underlying memory).

Reviewer 5: Thanks for this clarification! I was just wondering about this.
